pkg/cmd/cli/cmd/debug.go

```diff
@@ -0,0 +1,575 @@
+	if len(o.Filename) > 0 {
+		b.FilenameParam(explicit, o.Filename)
+	}
+
+	o.AddEnv, o.RemoveEnv, err = cmdutil.ParseEnv(envArgs, nil)
+	if err != nil {
+		return err
+	}
+
+	one := false
+	infos, err := b.Do().IntoSingular(&one).Infos()
+	if err != nil {
+		return err
+	}
+	if !one {
+		return fmt.Errorf("you must identify a resource with a pod template to debug")
+	}
+
+	template, err := f.ApproximatePodTemplateForObject(infos[0].Object)
+	if err != nil || template == nil {
+		return fmt.Errorf("cannot debug %s: %v", infos[0].Name, err)
+	}
+	pod := &kapi.Pod{
+		ObjectMeta: template.ObjectMeta,
+		Spec:       template.Spec,
+	}
+	pod.Name, pod.Namespace = infos[0].Name, infos[0].Namespace
+	o.Attach.Pod = pod
+
+	if len(o.Attach.ContainerName) == 0 && len(pod.Spec.Containers) > 0 {
+		glog.V(4).Infof("defaulting container name to %s", pod.Spec.Containers[0].Name)
+		o.Attach.ContainerName = pod.Spec.Containers[0].Name
+	}
+
+	o.Annotations[debugPodAnnotationSourceResource] = fmt.Sprintf("%s/%s", infos[0].Mapping.Resource, infos[0].Name)
+	o.Annotations[debugPodAnnotationSourceContainer] = o.Attach.ContainerName
+
+	output := kcmdutil.GetFlagString(cmd, "output")
+	if len(output) != 0 {
+		o.Print = func(pod *kapi.Pod, out io.Writer) error {
+			return f.PrintObject(cmd, pod, out)
+		}
+	}
+
+	config, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+	o.Attach.Config = config
+
+	_, kc, err := f.Clients()
+	if err != nil {
+		return err
+	}
+	o.Attach.Client = kc
+	return nil
+}
+func (o DebugOptions) Validate() error {
+	names := containerNames(o.Attach.Pod)
+	if len(names) == 0 {
+		return fmt.Errorf("the provided pod must have at least one container")
+	}
+	if len(o.Attach.ContainerName) == 0 {
+		return fmt.Errorf("you must provide a container name to debug")
+	}
+	if containerForName(o.Attach.Pod, o.Attach.ContainerName) == nil {
+		return fmt.Errorf("the container %q is not a valid container name; must be one of %v", o.Attach.ContainerName, names)
+	}
+	return nil
+}
+
+// WatchConditionFunc returns true if the condition has been reached, false if it has not been reached yet,
+// or an error if the condition cannot be checked and should terminate.
+type WatchConditionFunc func(event watch.Event) (bool, error)
+
+// Until reads items from the watch until each provided condition succeeds, and then returns the last watch
+// encountered. The first condition that returns an error terminates the watch (and the event is also returned).
+// If no event has been received, the returned event will be nil.
+// TODO: move to pkg/watch upstream
+func Until(timeout time.Duration, watcher watch.Interface, conditions ...WatchConditionFunc) (*watch.Event, error) {
+	ch := watcher.ResultChan()
+	defer watcher.Stop()
+	var after <-chan time.Time
+	if timeout > 0 {
+		after = time.After(timeout)
+	} else {
+		ch := make(chan time.Time)
+		close(ch)
+		after = ch
+	}
+	var lastEvent *watch.Event
+	for _, condition := range conditions {
```

Comment:
> Or do you expect conditions to be fullfilled sequentially?

Heh, I had the same argument.  I came down to this: it's easier to build a `UnionWatchCondition` that *doesn't* care about order than it is to build one that *does* care about order, so the we build the harder one into the library.